crates/astria-sequencer/src/transaction/mod.rs

```diff
@@ -1,9 +1,131 @@
-pub mod action;
 pub(crate) mod action_handler;
-pub mod signed;
-pub mod unsigned;
 
-pub use action::Action;
 pub(crate) use action_handler::ActionHandler;
-pub use signed::Signed;
-pub use unsigned::Unsigned;
+use anyhow::{
+    ensure,
+    Context as _,
+};
+use proto::native::sequencer::v1alpha1::{
+    Action,
+    Address,
+    SignedTransaction,
+    UnsignedTransaction,
+};
+use tracing::instrument;
+
+use crate::accounts::state_ext::{
+    StateReadExt,
+    StateWriteExt,
+};
+
+pub(crate) fn check_stateless(tx: &SignedTransaction) -> anyhow::Result<()> {
+    tx.unsigned_transaction().check_stateless()
+}
+
+pub(crate) async fn check_stateful<S: StateReadExt + 'static>(
+    tx: &SignedTransaction,
+    state: &S,
+) -> anyhow::Result<()> {
+    let signer_address = Address::from_verification_key(tx.verification_key());
+    tx.unsigned_transaction()
+        .check_stateful(state, signer_address)
+        .await
+}
+
+pub(crate) async fn execute<S: StateWriteExt>(
+    tx: &SignedTransaction,
+    state: &mut S,
+) -> anyhow::Result<()> {
+    let signer_address = Address::from_verification_key(tx.verification_key());
+    tx.unsigned_transaction()
+        .execute(state, signer_address)
+        .await
+}
```

Comment:
why not implement these on `SignedTransaction` (like it was before)?

Reply:
Because `ActionHandler::execute` takes a `signer_address`, but `SignedTransaction` is the thing that a) constructs the signer address and then b) calls `<UnsignedTransaction as ActionHandler>::execute` with it.

So implementing `ActionHandler for SignedTransaction` does not make sense.

On the other hand doing `impl SignedTransaction { fn execute(...) -> ... }` the way it was before doesn't work either because of coherence (`SignedTransaction` is an external type).

The only change I think might make sense is to change `ActionHandler` entirely to the definition below and `impl ActionHandler for SignedTransaction` and remove the impl for `UnsignedTransaction`:
```rust
trait ActionHandler {
    async fn execute<S: StateWriteExt>(&self, state: &mut S) -> Result<()> {
        todo!()
    }
}
```
But I didn't want to redesign it just that much

Reply:
that makes sense, I didn't necessarily mean `impl ActionHandler for SignedTransaction` but you could make these methods on `SignedTransaction`?

Reply:
Unfortunately you can only implement inherent implementations in the crate that owns the original definition (so in this case, `astria-proto`). But doing that there does not make sense because this is sequencer-private functionality.

I could make a single-use trait just to implement these methods on `SignedTransaction`. Or I could make a newtype wrapper for `SignedTransaction`. But both seem overkill.